File: javascript/src/utils/MemoryManager.ts
import fs from 'fs';
import path from 'path';
import pino from 'pino';
import {
  BaseChatMemory,
  ConversationBufferMemory,
  ConversationSummaryMemory,
  CombinedMemory,
  EntityMemory
} from 'langchain/memory';
import {
  AIMessage,
  HumanMessage,
  SystemMessage,
  BaseMessage
} from '@langchain/core/messages';
import { BaseLanguageModel } from '@langchain/core/language_models/base';
import { ChatMessageHistory } from 'langchain/stores/message/in_memory';

const logger = pino({
  transport: {
    target: 'pino-pretty',
    options: {
      colorize: true,
    },
  },
});

type MemoryStore = Record<string, Record<string, unknown>>;
type LangChainMemoryStore = Record<string, BaseChatMemory>;
type ConversationMessage = {
  role: string;
  content: string;
};

/**
 * Memory manager for maintaining context across agents and sessions.
 * Provides storage and retrieval of data for agents with support for
 * both in-memory and persistent storage and integration with LangChain.
 */
export class MemoryManager {
  private memoryStore: MemoryStore = {};
  private langchainMemories: LangChainMemoryStore = {};
  
  /**
   * Initialize the memory manager
   * 
   * @param persistToDisk - Whether to persist memory to disk
   * @param storageDir - Directory for persistent storage
   * @param llm - Optional language model for memory operations
   */
  constructor(
    private readonly persistToDisk: boolean = false,
    private readonly storageDir: string = './memory',
    private readonly llm?: BaseLanguageModel
  ) {
    if (persistToDisk) {
      try {
        if (!fs.existsSync(storageDir)) {
          fs.mkdirSync(storageDir, { recursive: true });
        }
        logger.info(`Memory manager initialized with persistence at ${storageDir}`);
      } catch (error) {
        logger.error(`Failed to create storage directory: ${error}`);
        throw new Error(`Failed to create storage directory: ${error}`);
      }
    } else {
      logger.info('Memory manager initialized with in-memory storage only');
    }
  }
  
  /**
   * Store a value in memory
   * 
   * @param agentId - The ID of the agent storing the data
   * @param key - The key to store the value under
   * @param value - The value to store
   */
  public store(agentId: string, key: string, value: unknown): void {
    if (!this.memoryStore[agentId]) {
      this.memoryStore[agentId] = {};
    }
    
    this.memoryStore[agentId][key] = value;
    logger.debug(`Stored value for agent ${agentId} under key '${key}'`);
    
    if (this.persistToDisk) {
      this.saveToDisk(agentId);
    }
  }
  
  /**
   * Retrieve a value from memory
   * 
   * @param agentId - The ID of the agent retrieving the data
   * @param key - The key to retrieve
   * @returns The stored value if found, undefined otherwise
   */
  public retrieve(agentId: string, key: string): unknown {
    // If we're using persistence and don't have this agent's data in memory, try to load it
    if (this.persistToDisk && !this.memoryStore[agentId]) {
      this.loadFromDisk(agentId);
    }
    
    if (this.memoryStore[agentId] && this.memoryStore[agentId][key] !== undefined) {
      logger.debug(`Retrieved value for agent ${agentId} under key '${key}'`);
      return this.memoryStore[agentId][key];
    }
    
    logger.debug(`No value found for agent ${agentId} under key '${key}'`);
    return undefined;
  }
  
  /**
   * Get all stored values for an agent
   * 
   * @param agentId - The ID of the agent
   * @returns Dictionary of all keys and values for the agent
   */
  public getAll(agentId: string): Record<string, unknown> {
    // If we're using persistence and don't have this agent's data in memory, try to load it
    if (this.persistToDisk && !this.memoryStore[agentId]) {
      this.loadFromDisk(agentId);
    }
    
    return this.memoryStore[agentId] || {};
  }
  
  /**
   * Clear stored memory
   * 
   * @param agentId - If provided, clear only this agent's data, otherwise clear all
   */
  public clear(agentId?: string): void {
    if (agentId) {
      if (this.memoryStore[agentId]) {
        delete this.memoryStore[agentId];
        logger.info(`Cleared memory for agent ${agentId}`);
        
        if (this.langchainMemories[agentId]) {
          delete this.langchainMemories[agentId];
          logger.info(`Cleared LangChain memory for agent ${agentId}`);
        }
        
        if (this.persistToDisk) {
          // Clear regular memory file
          const filePath = path.join(this.storageDir, `${agentId}.json`);
          if (fs.existsSync(filePath)) {
            fs.unlinkSync(filePath);
          }
          
          // Clear LangChain memory file
          const lcFilePath = path.join(this.storageDir, `${agentId}_langchain.json`);
          if (fs.existsSync(lcFilePath)) {
            fs.unlinkSync(lcFilePath);
          }
        }
      }
    } else {
      this.memoryStore = {};
      this.langchainMemories = {};
      logger.info('Cleared all memory');
      
      if (this.persistToDisk) {
        try {
          const files = fs.readdirSync(this.storageDir);
          for (const file of files) {
            if (file.endsWith('.json')) {
              fs.unlinkSync(path.join(this.storageDir, file));
            }
          }
        } catch (error) {
          logger.error(`Failed to clear disk storage: ${error}`);
        }
      }
    }
  }
  
  /**
   * Save agent data to disk
   * 
   * @param agentId - The ID of the agent
   */
  private saveToDisk(agentId: string): void {
    try {
      // Save regular memory
      const filePath = path.join(this.storageDir, `${agentId}.json`);
      fs.writeFileSync(filePath, JSON.stringify(this.memoryStore[agentId]));
      logger.debug(`Saved memory to disk for agent ${agentId}`);
      
      // Save LangChain memory if it exists
      if (this.langchainMemories[agentId]) {
        // We need to handle serialization of LangChain memory differently
        // First, extract the messages from the memory's chatHistory
        try {
          const lcFilePath = path.join(this.storageDir, `${agentId}_langchain.json`);
          const chatHistory = this.getSerializableChatHistory(agentId);
          fs.writeFileSync(lcFilePath, JSON.stringify(chatHistory));
          logger.debug(`Saved LangChain memory to disk for agent ${agentId}`);
        } catch (error) {
          logger.error(`Failed to save LangChain memory to disk for agent ${agentId}: ${error}`);
        }
      }
    } catch (error) {
      logger.error(`Failed to save memory to disk for agent ${agentId}: ${error}`);
    }
  }
  
  /**
   * Get serializable chat history from a LangChain memory
   * 
   * @param agentId - The ID of the agent
   * @returns Array of serializable messages
   */
  private getSerializableChatHistory(agentId: string): ConversationMessage[] {
    const memory = this.langchainMemories[agentId];
    if (!memory) return [];
    
    try {
      // Try to access chat_memory
      const chatMemory = (memory as any).chatMemory;
      if (chatMemory && Array.isArray(chatMemory.messages)) {
        return chatMemory.messages.map((msg: BaseMessage) => ({
          role: this.getMessageRole(msg),
          content: msg.content
        }));
      }
      
      // Try getting memory variables
      if (typeof memory.loadMemoryVariables === 'function') {
        const vars = memory.loadMemoryVariables({});
        // Check if it's a Promise and handle accordingly
        if (vars instanceof Promise) {
          // We can't do async here, so we'll have to fall back to a stored history
          const storedHistory = this.retrieve(agentId, 'conversation_history');
          if (Array.isArray(storedHistory)) {
            return storedHistory as ConversationMessage[];
          }
          return [];
        }
        
        // If it's the actual variables, try to extract messages
        if (vars.chat_history && Array.isArray(vars.chat_history)) {
          return vars.chat_history.map((msg: BaseMessage) => ({
            role: this.getMessageRole(msg),
            content: msg.content
          }));
        }
      }
      
      // Fall back to stored history
      const storedHistory = this.retrieve(agentId, 'conversation_history');
      if (Array.isArray(storedHistory)) {
        return storedHistory as ConversationMessage[];
      }
      
      return [];
    } catch (error) {
      logger.error(`Error getting serializable chat history: ${error}`);
      return [];
    }
  }
  
  /**
   * Get the role from a LangChain message
   * 
   * @param message - A LangChain message
   * @returns The message role as a string
   */
  private getMessageRole(message: BaseMessage): string {
    if (message instanceof HumanMessage) {
      return 'user';
    } else if (message instanceof AIMessage) {
      return 'assistant';
    } else if (message instanceof SystemMessage) {
      return 'system';
    } else {
      return 'unknown';
    }
  }
  
  /**
   * Load agent data from disk
   * 
   * @param agentId - The ID of the agent
   */
  private loadFromDisk(agentId: string): void {
    // Load regular memory
    const filePath = path.join(this.storageDir, `${agentId}.json`);
    if (fs.existsSync(filePath)) {
      try {
        const fileContent = fs.readFileSync(filePath, 'utf8');
        this.memoryStore[agentId] = JSON.parse(fileContent);
        logger.debug(`Loaded memory from disk for agent ${agentId}`);
      } catch (error) {
        logger.error(`Failed to load memory from disk for agent ${agentId}: ${error}`);
        // Initialize empty object if load fails
        this.memoryStore[agentId] = {};
      }
    }
    
    // Load LangChain memory if it exists
    const lcFilePath = path.join(this.storageDir, `${agentId}_langchain.json`);
    if (fs.existsSync(lcFilePath)) {
      try {
        const fileContent = fs.readFileSync(lcFilePath, 'utf8');
        const chatHistory = JSON.parse(fileContent) as ConversationMessage[];
        
        // We'll store this in the regular memory for now
        // and create a proper LangChain memory when getLangChainMemory is called
        this.store(agentId, 'conversation_history', chatHistory);
        logger.debug(`Loaded serialized LangChain memory for agent ${agentId}`);
      } catch (error) {
        logger.error(`Failed to load LangChain memory from disk for agent ${agentId}: ${error}`);
      }
    }
  }
  
  // LangChain Memory Integration
  
  /**
   * Get a LangChain memory for an agent
   * 
   * @param agentId - The ID of the agent
   * @param memoryType - Type of memory to use ('buffer', 'summary', 'entity', or 'combined')
   * @param memoryKey - Key for the memory in output context
   * @returns A LangChain memory instance
   */
  public getLangChainMemory(
    agentId: string, 
    memoryType: 'buffer' | 'summary' | 'entity' | 'combined' = 'buffer',
    memoryKey: string = 'chat_history'
  ): BaseChatMemory {
    // Check if we already have a memory for this agent
    if (this.langchainMemories[agentId]) {
      return this.langchainMemories[agentId];
    }
    
    // Create a new memory based on the type
    let memory: BaseChatMemory;
    
    if (memoryType === 'buffer') {
      memory = new ConversationBufferMemory({
        memoryKey: memoryKey,
        returnMessages: true
      });
    } else if (memoryType === 'summary') {
      if (!this.llm) {
        logger.warning('No LLM provided for summary memory, falling back to buffer memory');
        memory = new ConversationBufferMemory({
          memoryKey: memoryKey,
          returnMessages: true
        });
      } else {
        memory = new ConversationSummaryMemory({
          llm: this.llm,
          memoryKey: memoryKey,
          returnMessages: true
        });
      }
    } else if (memoryType === 'entity') {
      if (!this.llm) {
        logger.warning('No LLM provided for entity memory, falling back to buffer memory');
        memory = new ConversationBufferMemory({
          memoryKey: memoryKey,
          returnMessages: true
        });
      } else {
        memory = new EntityMemory({
          llm: this.llm,
          entityCache: {},
          entitiesKey: 'entities',
          chatHistoryKey: memoryKey
        });
      }
    } else if (memoryType === 'combined') {
      if (!this.llm) {
        logger.warning('No LLM provided for combined memory, falling back to buffer memory');
        memory = new ConversationBufferMemory({
          memoryKey: memoryKey,
          returnMessages: true
        });
      } else {
        const bufferMemory = new ConversationBufferMemory({
          memoryKey: 'chat_history',
          returnMessages: true
        });
        const summaryMemory = new ConversationSummaryMemory({
          llm: this.llm,
          memoryKey: 'summary_history',
          returnMessages: true
        });
        memory = new CombinedMemory({
          memories: [bufferMemory, summaryMemory],
          chatHistory: new ChatMessageHistory()
        });
      }
    } else {
      // Default to buffer memory
      memory = new ConversationBufferMemory({
        memoryKey: memoryKey,
        returnMessages: true
      });
    }
    
    // Load any existing chat history from regular memory
    const existingHistory = this.retrieve(agentId, 'conversation_history');
    if (existingHistory && Array.isArray(existingHistory)) {
      try {
        // Initialize the chat history with stored messages
        const chatHistory = existingHistory as ConversationMessage[];
        for (const msg of chatHistory) {
          if (msg.role === 'user') {
            memory.chatHistory.addUserMessage(msg.content);
          } else if (msg.role === 'assistant') {
            memory.chatHistory.addAIMessage(msg.content);
          }
          // System messages typically aren't stored in chat history
        }
        logger.debug(`Loaded ${chatHistory.length} messages into LangChain memory for agent ${agentId}`);
      } catch (error) {
        logger.error(`Error loading chat history into LangChain memory: ${error}`);
      }
    }
    
    // Store the memory
    this.langchainMemories[agentId] = memory;
    logger.info(`Created new ${memoryType} memory for agent ${agentId}`);
    
    // Persist if needed
    if (this.persistToDisk) {
      this.saveToDisk(agentId);
    }
    
    return memory;
  }
  
  /**
   * Save a message to the agent's LangChain memory
   * 
   * @param agentId - The ID of the agent
   * @param message - The message content
   * @param role - The role of the message sender ('user', 'assistant', 'system')
   */
  public saveMessageToMemory(agentId: string, message: string, role: string): void {
    // Get or create the agent's memory
    const memory = this.getLangChainMemory(agentId);
    
    // Add message to regular memory store too
    let conversationHistory = this.retrieve(agentId, 'conversation_history') as ConversationMessage[];
    if (!conversationHistory || !Array.isArray(conversationHistory)) {
      conversationHistory = [];
    }
    
    conversationHistory.push({ role, content: message });
    this.store(agentId, 'conversation_history', conversationHistory);
    
    // Add the message to the LangChain memory
    try {
      if (role === 'user') {
        memory.chatHistory.addUserMessage(message);
      } else if (role === 'assistant') {
        memory.chatHistory.addAIMessage(message);
      } else if (role === 'system') {
        // Some memory implementations don't support system messages directly
        // Store system messages in a separate key
        const timestamp = new Date().toISOString();
        this.store(agentId, `system_message_${timestamp}`, message);
      }
      
      logger.debug(`Added ${role} message to LangChain memory for agent ${agentId}`);
      
      // Persist if needed
      if (this.persistToDisk) {
        this.saveToDisk(agentId);
      }
    } catch (error) {
      logger.error(`Error saving message to LangChain memory: ${error}`);
    }
  }
  
  /**
   * Get the conversation history for an agent
   * 
   * @param agentId - The ID of the agent
   * @param asMessages - Whether to return as LangChain messages or serialized
   * @returns Conversation history as messages or serialized objects
   */
  public async getConversationHistory(
    agentId: string, 
    asMessages: boolean = false
  ): Promise<BaseMessage[] | ConversationMessage[]> {
    try {
      // First try to get from LangChain memory
      if (this.langchainMemories[agentId]) {
        const memory = this.langchainMemories[agentId];
        
        if (memory.chatHistory) {
          const messages = memory.chatHistory.getMessages();
          
          if (asMessages) {
            return messages;
          } else {
            // Convert to serializable format
            return messages.map(msg => ({
              role: this.getMessageRole(msg),
              content: msg.content
            }));
          }
        }
        
        // Try to load from memory variables
        if (typeof memory.loadMemoryVariables === 'function') {
          try {
            const vars = await memory.loadMemoryVariables({});
            if (vars.chat_history && Array.isArray(vars.chat_history)) {
              if (asMessages) {
                return vars.chat_history;
              } else {
                return vars.chat_history.map(msg => ({
                  role: this.getMessageRole(msg),
                  content: msg.content
                }));
              }
            }
          } catch (error) {
            logger.error(`Error loading memory variables: ${error}`);
          }
        }
      }
      
      // Fall back to regular memory
      const history = this.retrieve(agentId, 'conversation_history') as ConversationMessage[];
      if (history && Array.isArray(history)) {
        if (asMessages) {
          // Convert to LangChain message format
          return history.map(msg => {
            if (msg.role === 'user') {
              return new HumanMessage(msg.content);
            } else if (msg.role === 'assistant') {
              return new AIMessage(msg.content);
            } else if (msg.role === 'system') {
              return new SystemMessage(msg.content);
            } else {
              // Default to human message for unknown roles
              return new HumanMessage(msg.content);
            }
          });
        } else {
          return history;
        }
      }
      
      // Return empty history if nothing found
      return [];
    } catch (error) {
      logger.error(`Error getting conversation history: ${error}`);
      return [];
    }
  }
}

File: python/src/utils/memory_manager.py
from typing import Any, Dict, Optional, List, Tuple, Type, cast
from loguru import logger
import json
import os
import pickle
import datetime

from langchain_core.memory import BaseMemory
from langchain.memory import (
    ConversationBufferMemory,
    ConversationSummaryMemory,
    CombinedMemory,
    ConversationEntityMemory
)
from langchain_core.messages import BaseMessage, AIMessage, HumanMessage, SystemMessage
from langchain_core.language_models import BaseLanguageModel

# BaseEntityStore was moved in newer versions of langchain
try:
    from langchain_core.message_history import BaseEntityStore
except ImportError:
    try:
        from langchain_core.entity_stores import BaseEntityStore
    except ImportError:
        # If both imports fail, create a stub class
        class BaseEntityStore:
            pass


class MemoryManager:
    """Memory manager for maintaining context across agents and sessions.
    
    This class provides storage and retrieval of data for agents,
    with support for both in-memory and persistent storage, and
    integration with LangChain memory components.
    """
    
    def __init__(self, persist_to_disk: bool = False, storage_dir: str = "./memory", 
                llm: Optional[BaseLanguageModel] = None):
        """Initialize the memory manager.
        
        Args:
            persist_to_disk: Whether to persist memory to disk
            storage_dir: Directory for persistent storage
            llm: Optional language model for memory operations
        """
        self._memory_store: Dict[str, Dict[str, Any]] = {}
        self._langchain_memories: Dict[str, BaseMemory] = {}
        self.persist_to_disk = persist_to_disk
        self.storage_dir = storage_dir
        self.llm = llm
        
        if persist_to_disk:
            os.makedirs(storage_dir, exist_ok=True)
            logger.info(f"Memory manager initialized with persistence at {storage_dir}")
        else:
            logger.info("Memory manager initialized with in-memory storage only")
    
    def store(self, agent_id: str, key: str, value: Any) -> None:
        """Store a value in memory.
        
        Args:
            agent_id: The ID of the agent storing the data
            key: The key to store the value under
            value: The value to store
        """
        if agent_id not in self._memory_store:
            self._memory_store[agent_id] = {}
        
        self._memory_store[agent_id][key] = value
        logger.debug(f"Stored value for agent {agent_id} under key '{key}'")
        
        if self.persist_to_disk:
            self._save_to_disk(agent_id)
    
    def retrieve(self, agent_id: str, key: str) -> Optional[Any]:
        """Retrieve a value from memory.
        
        Args:
            agent_id: The ID of the agent retrieving the data
            key: The key to retrieve
            
        Returns:
            The stored value if found, None otherwise
        """
        # If we're using persistence and don't have this agent's data in memory, try to load it
        if self.persist_to_disk and agent_id not in self._memory_store:
            self._load_from_disk(agent_id)
        
        if agent_id in self._memory_store and key in self._memory_store[agent_id]:
            logger.debug(f"Retrieved value for agent {agent_id} under key '{key}'")
            return self._memory_store[agent_id][key]
        
        logger.debug(f"No value found for agent {agent_id} under key '{key}'")
        return None
    
    def get_all(self, agent_id: str) -> Dict[str, Any]:
        """Get all stored values for an agent.
        
        Args:
            agent_id: The ID of the agent
            
        Returns:
            Dictionary of all keys and values for the agent
        """
        # If we're using persistence and don't have this agent's data in memory, try to load it
        if self.persist_to_disk and agent_id not in self._memory_store:
            self._load_from_disk(agent_id)
        
        return self._memory_store.get(agent_id, {})
    
    def clear(self, agent_id: Optional[str] = None) -> None:
        """Clear stored memory.
        
        Args:
            agent_id: If provided, clear only this agent's data, otherwise clear all
        """
        if agent_id:
            if agent_id in self._memory_store:
                del self._memory_store[agent_id]
                logger.info(f"Cleared memory for agent {agent_id}")
                
                if agent_id in self._langchain_memories:
                    del self._langchain_memories[agent_id]
                    logger.info(f"Cleared LangChain memory for agent {agent_id}")
                
                if self.persist_to_disk:
                    file_path = os.path.join(self.storage_dir, f"{agent_id}.json")
                    if os.path.exists(file_path):
                        os.remove(file_path)
                    
                    # Also remove LangChain memory files
                    lc_file_path = os.path.join(self.storage_dir, f"{agent_id}_langchain.pkl")
                    if os.path.exists(lc_file_path):
                        os.remove(lc_file_path)
        else:
            self._memory_store.clear()
            self._langchain_memories.clear()
            logger.info("Cleared all memory")
            
            if self.persist_to_disk:
                for filename in os.listdir(self.storage_dir):
                    if filename.endswith(('.json', '.pkl')):
                        os.remove(os.path.join(self.storage_dir, filename))
    
    def _save_to_disk(self, agent_id: str) -> None:
        """Save agent data to disk.
        
        Args:
            agent_id: The ID of the agent
        """
        try:
            file_path = os.path.join(self.storage_dir, f"{agent_id}.json")
            with open(file_path, 'w') as f:
                json.dump(self._memory_store[agent_id], f)
            logger.debug(f"Saved memory to disk for agent {agent_id}")
            
            # Also save LangChain memory if it exists
            if agent_id in self._langchain_memories:
                lc_file_path = os.path.join(self.storage_dir, f"{agent_id}_langchain.pkl")
                with open(lc_file_path, 'wb') as f:
                    pickle.dump(self._langchain_memories[agent_id], f)
                logger.debug(f"Saved LangChain memory to disk for agent {agent_id}")
        except Exception as e:
            logger.error(f"Failed to save memory to disk for agent {agent_id}: {str(e)}")
    
    def _load_from_disk(self, agent_id: str) -> None:
        """Load agent data from disk.
        
        Args:
            agent_id: The ID of the agent
        """
        file_path = os.path.join(self.storage_dir, f"{agent_id}.json")
        if os.path.exists(file_path):
            try:
                with open(file_path, 'r') as f:
                    self._memory_store[agent_id] = json.load(f)
                logger.debug(f"Loaded memory from disk for agent {agent_id}")
            except Exception as e:
                logger.error(f"Failed to load memory from disk for agent {agent_id}: {str(e)}")
                # Initialize empty dict if load fails
                self._memory_store[agent_id] = {}
        
        # Also load LangChain memory if it exists
        lc_file_path = os.path.join(self.storage_dir, f"{agent_id}_langchain.pkl")
        if os.path.exists(lc_file_path):
            try:
                with open(lc_file_path, 'rb') as f:
                    self._langchain_memories[agent_id] = pickle.load(f)
                logger.debug(f"Loaded LangChain memory from disk for agent {agent_id}")
            except Exception as e:
                logger.error(f"Failed to load LangChain memory from disk for agent {agent_id}: {str(e)}")
                # Will recreate memory when needed
    
    # LangChain Memory Integration
    
    def get_langchain_memory(self, agent_id: str, memory_type: str = "buffer", 
                           memory_key: str = "chat_history") -> BaseMemory:
        """Get a LangChain memory for an agent.
        
        Args:
            agent_id: The ID of the agent
            memory_type: Type of memory to use ('buffer', 'summary', 'entity', or 'combined')
            memory_key: Key for the memory in output context
            
        Returns:
            A LangChain memory instance
        """
        # Check if we already have a memory for this agent
        if agent_id in self._langchain_memories:
            return self._langchain_memories[agent_id]
        
        # If we're using persistence, try to load the memory
        if self.persist_to_disk:
            lc_file_path = os.path.join(self.storage_dir, f"{agent_id}_langchain.pkl")
            if os.path.exists(lc_file_path):
                try:
                    with open(lc_file_path, 'rb') as f:
                        self._langchain_memories[agent_id] = pickle.load(f)
                    logger.debug(f"Loaded LangChain memory from disk for agent {agent_id}")
                    return self._langchain_memories[agent_id]
                except Exception as e:
                    logger.error(f"Failed to load LangChain memory from disk for agent {agent_id}: {str(e)}")
        
        # Create a new memory based on the type
        memory: BaseMemory
        
        if memory_type == "buffer":
            memory = ConversationBufferMemory(memory_key=memory_key, return_messages=True)
        elif memory_type == "summary":
            if not self.llm:
                logger.warning("No LLM provided for summary memory, falling back to buffer memory")
                memory = ConversationBufferMemory(memory_key=memory_key, return_messages=True)
            else:
                memory = ConversationSummaryMemory(
                    llm=self.llm,
                    memory_key=memory_key,
                    return_messages=True
                )
        elif memory_type == "entity":
            if not self.llm:
                logger.warning("No LLM provided for entity memory, falling back to buffer memory")
                memory = ConversationBufferMemory(memory_key=memory_key, return_messages=True)
            else:
                memory = ConversationEntityMemory(
                    llm=self.llm,
                    entity_cache={},
                    memory_key=memory_key,
                    return_messages=True,
                    verbose=True
                )
        elif memory_type == "combined":
            if not self.llm:
                logger.warning("No LLM provided for combined memory, falling back to buffer memory")
                memory = ConversationBufferMemory(memory_key=memory_key, return_messages=True)
            else:
                buffer_memory = ConversationBufferMemory(memory_key="chat_history", return_messages=True)
                summary_memory = ConversationSummaryMemory(
                    llm=self.llm,
                    memory_key="summary_history",
                    return_messages=True
                )
                memory = CombinedMemory(memories=[buffer_memory, summary_memory])
        else:
            logger.warning(f"Unknown memory type {memory_type}, using buffer memory")
            memory = ConversationBufferMemory(memory_key=memory_key, return_messages=True)
        
        # Store the memory
        self._langchain_memories[agent_id] = memory
        logger.info(f"Created new {memory_type} memory for agent {agent_id}")
        
        # Persist if needed
        if self.persist_to_disk:
            self._save_to_disk(agent_id)
        
        return memory
    
    def save_message_to_memory(self, agent_id: str, message: str, role: str) -> None:
        """Save a message to the agent's LangChain memory.
        
        Args:
            agent_id: The ID of the agent
            message: The message content
            role: The role of the message sender ('user', 'assistant', 'system')
        """
        # Get or create the agent's memory
        memory = self.get_langchain_memory(agent_id)
        
        # Add the message to the memory
        chat_memory = getattr(memory, "chat_memory", None)
        if chat_memory:
            if role == "user":
                chat_memory.add_user_message(message)
            elif role == "assistant":
                chat_memory.add_ai_message(message)
            elif role == "system":
                # Some memory implementations don't support system messages directly
                # Store it as a key in the regular memory
                self.store(agent_id, f"system_message_{datetime.datetime.now().isoformat()}", message)
            
            logger.debug(f"Added {role} message to LangChain memory for agent {agent_id}")
            
            # Persist if needed
            if self.persist_to_disk:
                self._save_to_disk(agent_id)
        else:
            logger.warning(f"Could not add message to LangChain memory for agent {agent_id}: no chat_memory attribute")
    
    def get_conversation_history(self, agent_id: str, as_messages: bool = False) -> Any:
        """Get the conversation history for an agent.
        
        Args:
            agent_id: The ID of the agent
            as_messages: Whether to return as LangChain messages or text
            
        Returns:
            Conversation history as messages or text
        """
        # Get the agent's memory
        memory = self.get_langchain_memory(agent_id)
        
        try:
            # Try to get the variables (different memory implementations have different methods)
            if hasattr(memory, "load_memory_variables"):
                memory_data = memory.load_memory_variables({})
                if as_messages:
                    # Return messages directly if available
                    if isinstance(memory_data.get("chat_history", []), List):
                        return memory_data.get("chat_history", [])
                    # Try to get chat_memory messages
                    chat_memory = getattr(memory, "chat_memory", None)
                    if chat_memory and hasattr(chat_memory, "messages"):
                        return chat_memory.messages
                    
                # Return as string
                if "chat_history" in memory_data:
                    return memory_data["chat_history"]
                # Try alternate keys
                for key in memory_data:
                    if "history" in key or "memory" in key:
                        return memory_data[key]
            
            # Fallback to our stored conversation history
            legacy_history = self.retrieve(agent_id, "conversation_history")
            if legacy_history and isinstance(legacy_history, list):
                if as_messages:
                    # Convert dict format to LangChain messages
                    messages = []
                    for msg in legacy_history:
                        if msg["role"] == "user":
                            messages.append(HumanMessage(content=msg["content"]))
                        elif msg["role"] == "assistant":
                            messages.append(AIMessage(content=msg["content"]))
                        elif msg["role"] == "system":
                            messages.append(SystemMessage(content=msg["content"]))
                    return messages
                else:
                    # Format as text
                    return "\n".join([f"{msg['role']}: {msg['content']}" for msg in legacy_history])
            
            # No history found
            return [] if as_messages else ""
            
        except Exception as e:
            logger.error(f"Error retrieving conversation history for agent {agent_id}: {str(e)}")
            return [] if as_messages else ""